common/src/main/java/org/apache/celeborn/common/network/sasl/registration/RegistrationClientBootstrap.java

```diff
@@ -0,0 +1,260 @@
+import org.apache.celeborn.common.network.client.TransportClientBootstrap;
+import org.apache.celeborn.common.network.protocol.TransportMessage;
+import org.apache.celeborn.common.network.sasl.CelebornSaslClient;
+import org.apache.celeborn.common.network.sasl.SaslClientBootstrap;
+import org.apache.celeborn.common.network.sasl.SaslCredentials;
+import org.apache.celeborn.common.network.sasl.SaslTimeoutException;
+import org.apache.celeborn.common.network.util.TransportConf;
+import org.apache.celeborn.common.protocol.MessageType;
+import org.apache.celeborn.common.protocol.PbAuthType;
+import org.apache.celeborn.common.protocol.PbAuthenticationInitiationRequest;
+import org.apache.celeborn.common.protocol.PbAuthenticationInitiationResponse;
+import org.apache.celeborn.common.protocol.PbRegisterApplicationRequest;
+import org.apache.celeborn.common.protocol.PbRegisterApplicationResponse;
+import org.apache.celeborn.common.protocol.PbSaslMechanism;
+import org.apache.celeborn.common.protocol.PbSaslRequest;
+import org.apache.celeborn.common.util.JavaUtils;
+
+/**
+ * Bootstraps a {@link TransportClient} by registering application (if the application is not
+ * registered). If the application is already registered, it will bootstrap the client by performing
+ * SASL authentication.
+ */
+public class RegistrationClientBootstrap implements TransportClientBootstrap {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RegistrationClientBootstrap.class);
+
+  private static final String VERSION = "1.0";
+
+  /**
+   * TODO: This should be made configurable. For now, we only support ANONYMOUS for client-auth and
+   * DIGEST-MD5 for connect-auth.
+   */
+  private static final List<PbSaslMechanism> SASL_MECHANISMS =
+      Lists.newArrayList(
+          PbSaslMechanism.newBuilder()
+              .setMechanism(ANONYMOUS)
+              .addAuthTypes(PbAuthType.CLIENT_AUTH)
+              .build(),
+          PbSaslMechanism.newBuilder()
+              .setMechanism(DIGEST_MD5)
+              .addAuthTypes(PbAuthType.CONNECTION_AUTH)
+              .build());
+
+  private final TransportConf conf;
+  private final String appId;
+  private final SaslCredentials saslCredentials;
+
+  private final RegistrationInfo registrationInfo;
+
+  public RegistrationClientBootstrap(
+      TransportConf conf,
+      String appId,
+      SaslCredentials saslCredentials,
+      RegistrationInfo registrationInfo) {
+    this.conf = Preconditions.checkNotNull(conf, "conf");
+    this.appId = Preconditions.checkNotNull(appId, "appId");
+    this.saslCredentials = Preconditions.checkNotNull(saslCredentials, "saslCredentials");
+    this.registrationInfo = Preconditions.checkNotNull(registrationInfo, "registrationInfo");
+  }
+
+  @Override
+  public void doBootstrap(TransportClient client, Channel channel) throws RuntimeException {
+    if (registrationInfo.getRegistrationState() == RegistrationInfo.RegistrationState.REGISTERED) {
```

Comment:
To clarify, we are allowing for repeated re-registeration if the previous requests failed; is this expected behavior we are shooting for ? Or will a failure result in application termination ?

Reply:
We are not allowing for re-registeration if the previous registeration failed in the current implementation. At the server it checks when it gets the AuthenticationInitiaionRequest that the registration state should be `NONE`.
    ```
        PbAuthenticationInitiationRequest authInitRequest = pbMsg.getParsedPayload();
        checkRequestAllowed(RegistrationState.NONE);
        respondToAuthInitialization(callback);
      ```
I do think, I should add a `FAILED` state at the server and set it when registration fails.